ext/bcmath/bcmath.c

```diff
@@ -777,4 +791,965 @@ PHP_FUNCTION(bcscale)
+		return SUCCESS;
+	} else if (str) {
+		return bc_num_from_zend_string(str, num, full_scale);
+	} else {
+		bc_num_from_zend_long(lval, num, full_scale);
+		return SUCCESS;
+	}
+}
+
+static zend_result bcmath_number_do_operation(uint8_t opcode, zval *ret_val, zval *op1, zval *op2)
+{
+	switch (opcode) {
+		case ZEND_ADD:
+		case ZEND_SUB:
+		case ZEND_MUL:
+		case ZEND_DIV:
+		case ZEND_MOD:
+		case ZEND_POW:
+			break;
+		default:
+			return FAILURE;
+	}
+
+	zend_object *obj1 = NULL;
+	zend_string *str1 = NULL;
+	zend_long lval1;
+
+	zend_object *obj2 = NULL;
+	zend_string *str2 = NULL;
+	zend_long lval2;
+
+	if (UNEXPECTED(Z_TYPE_P(op1) == IS_NULL)) {
+		lval1 = 0;
+	} else if (UNEXPECTED(bcmath_number_parse_num(op1, &obj1, &str1, &lval1) == FAILURE)) {
+		return FAILURE;
+	}
+
+	if (UNEXPECTED(Z_TYPE_P(op2) == IS_NULL)) {
+		lval2 = 0;
+	} else if (UNEXPECTED(bcmath_number_parse_num(op2, &obj2, &str2, &lval2) == FAILURE)) {
+		return FAILURE;
+	}
+
+	bc_num n1 = NULL;
+	bc_num n2 = NULL;
+	size_t n1_full_scale;
+	size_t n2_full_scale;
+	if (UNEXPECTED(bc_num_from_obj_or_str_or_long(&n1, &n1_full_scale, obj1, str1, lval1) == FAILURE ||
+		bc_num_from_obj_or_str_or_long(&n2, &n2_full_scale, obj2, str2, lval2) == FAILURE)) {
+		zend_value_error("Number is not well-formed");
+		goto fail;
+	}
+
+	/* For increment and decrement, etc */
+	zval op1_copy;
+	if (ret_val == op1) {
+		ZVAL_COPY_VALUE(&op1_copy, op1);
+		op1 = &op1_copy;
+	}
+
+	bc_num ret = NULL;
+	size_t scale;
+	switch (opcode) {
+		case ZEND_ADD:
+			bcmath_number_add_internal(n1, n2, &ret, n1_full_scale, n2_full_scale, &scale, true);
+			break;
+		case ZEND_SUB:
+			bcmath_number_sub_internal(n1, n2, &ret, n1_full_scale, n2_full_scale, &scale, true);
+			break;
+		case ZEND_MUL:
+			bcmath_number_mul_internal(n1, n2, &ret, n1_full_scale, n2_full_scale, &scale, true);
+			break;
+		case ZEND_DIV:
+			if (UNEXPECTED(bcmath_number_div_internal(n1, n2, &ret, n1_full_scale, n2_full_scale, &scale, true) == FAILURE)) {
+				goto fail;
+			}
+			break;
+		case ZEND_MOD:
+			if (UNEXPECTED(bcmath_number_mod_internal(n1, n2, &ret, n1_full_scale, n2_full_scale, &scale, true) == FAILURE)) {
+				goto fail;
+			}
+			break;
+		case ZEND_POW:
+			if (UNEXPECTED(bcmath_number_pow_internal(n1, n2, &ret, n1_full_scale, n2_full_scale, &scale, true, true) == FAILURE)) {
+				goto fail;
+			}
+			break;
+		EMPTY_SWITCH_DEFAULT_CASE();
+	}
+
+	if (Z_TYPE_P(op1) != IS_OBJECT) {
+		bc_free_num(&n1);
+	}
+	if (Z_TYPE_P(op2) != IS_OBJECT) {
+		bc_free_num(&n2);
+	}
+
+	bcmath_number_obj_t *intern = bcmath_number_new_obj(ret, scale);
+	ZVAL_OBJ(ret_val, &intern->std);
+
+	/* For increment and decrement, etc */
+	if (op1 == &op1_copy) {
+		zval_ptr_dtor(op1);
+	}
+
+	return SUCCESS;
+
+fail:
+	if (Z_TYPE_P(op1) != IS_OBJECT) {
+		bc_free_num(&n1);
+	}
+	if (Z_TYPE_P(op2) != IS_OBJECT) {
+		bc_free_num(&n2);
+	}
+	return FAILURE;
+}
+
+static int bcmath_number_compare(zval *op1, zval *op2)
```

Comment:
Parameter names do not match the prototype.

Reply:
Fixed in https://github.com/php/php-src/pull/13741/commits/af44a39ca66f94667e41761878c30fb36cf71362